### DISCUSSION TITLE
Zu Listing 1.16 - Gemeinsames Profil für "Windows PowerShell" und "PowerShell"

### DISCUSSION THREAD

**Discussion by helmutkilian** [TobiasPSP/OReilly#10] [Ideen]
Die gute Idee und Methode (Hardlink) aus dem Buch, ein gemeinsames Profil einzurichten, habe ich sogleich für mich umgesetzt!

Extra bemerkenswert ist, dass man das hierfür bereitgestellte Skript 1.16 nach Belieben unter "Windows Powershell" oder "Powershell" ausführen kann, um dasselbe Ergebnis zu erreichen. 

**Frage:** Gibt es einen Grund, warum das Skript für "Windows Powershell" die Profildatei "$profile.CurrentUserAllHosts" benutzt und für "Powershell" die Profildatei "$profile.CurrentUserCurrentHost" (was dasselbe ist wie "$profile")?

Ich sehe keinen Grund. Für mich habe ich deshalb in Zeilen 17 und 23 von Listing 1.16 
`$coreprofile =  $profile`
analog zu $winprofile durch 
`$coreprofile =  $profile.CurrentUserAllHosts`
ersetzt.

Unabhängig davon gab es ein **Problem**: Ich hatte den Hardlink mit dem Skript erzeugt und dann durch Löschen der Profildatei für "Windows Powershell" probeweise wieder aufgehoben. Danach ließ sich der Hardlink mit Skript 1.16 **nicht erneut herstellen**.
Die Ursache habe ich gefunden. Durch einfaches Löschen einer der beiden Dateien im Windows-Explorer bleibt bei der anderen Datei die Eigenschaft LinkType = 'HardLink' erhalten. Durch "einfaches" Löschen wird die "gelöschte" Datei nur in den Papierkorb verschoben, wobei der Hardlink erhalten bleibt.

Lösung: Die Warnung in Zeile 77 des Skripts habe ich für mich wie folgt geändert und die Änderung kommentiert:

`Write-Warning "Um die Spiegelung aufzuheben, eine der beiden Profildateien, $coreprofile oder $winprofile, unwiderruflich löschen (z. B. im Windows-Explorer mit Shift+Entf) und neu angelegen."`
`# Hinweis zum Löschen: Verschieben in den Papierkorb genügt nicht, um den Hardlink zu entfernen. Die verbleibende Datei würde die Eingenschaft LinkType mit dem Wert 'HardLink' behalten, und dieses Skript würde bei einem erneuten Verlinkungsversuch falsch reagieren.`

Anmerkung: Würde man die Datei "$coreprofile" statt "$winprofile" in den Papierkorb verschieben, so würde das Skript erfolgreich reagieren. Es gäbe dann aber drei Orte der gemeinsame Profildatei:  "$coreprofile", "$winprofile" und Papierkorb.

**Comment by Federahannes** [comment]
Habe das Listing bei mir auch jeweils für PS bzw. Windows PS erfolgreich implementiert. In VSCode war der L Befehl zunächst aber nicht verfügbar. Musste das Prozedere in dem dortigen Terminal auch nochmal durchführen (obwohl diese Profildatei als sie "aufging" bereits die Zeilen enthielt wie ich sie für PS bzw. WPS gespeichert hatte).

**Reply by TobiasPSP** [reply]
Kann es sein, dass die PowerShell in VSCode die Autostartdatei in $profile nicht automatisch ausgeführt hat, z.B. weil die ExecutionPolicy noch auf Restricted/Default stand? 

Das Skript, dessen Pfadname innerhalb der PowerShell in $profile steht, wird in jedem Fall automatisch beim Start ausgeführt - nur falls die ExecutionPolicy die Skriptausführung nicht gestattet, gilt dies auch für Profilskripte, so dass nur dann der automatische Start unterbleibt. Wenn also in dieser Datei die Funktion angelegt wurde, sollte sie eigentlich nach dem Start automatisch vorhanden sein.

**Reply by Federahannes** [reply]
Der Laptop ist vom Bildungsträger gemanged. Die ExecutionPolicies hatte ich überall (VSC, PS, WPS) für dies Scopes Process, CurrentUser und LocalMachine jeweils auf ByPass gesetzt, aber die Einstellung für "Process" geht nach jedem Neustart zurück auf "Undefined". (mehr kann ich leider nicht dazu sagen, da ich was VSCode - und Windows überhaupt - total naiv bin )

**Comment by TobiasPSP** [comment]
Helmut, danke für Deine detaillierten Anmerkungen. Bei mir sind leider Benachrichtigungen ausgefallen, weswegen ich Deinen Kommentar erst jetzt entdeckt habe.
Der Unterschied zwischen "$profile.CurrentUserAllHosts" und "$profile.CurrentUserCurrentHost" ist der Host, d.h. ein Profilskript "$profile.CurrentUserCurrentHost" läuft nur in dem Host, in dem es definiert wurde (also z.B. der Konsole). 

"$profile" liefert by default dieselbe Einstellung wie "$profile.CurrentUserCurrentHost". Da das Ziel war, eine (Default)Profildatei für beide PowerShell-Arten zu haben, wurde "$profile.CurrentUserCurrentHost" gespiegelt. Es kann problematisch sein, in seiner Profildatei Dinge automatisch auszuführen, die dann in einem völlig anderen Host (z.B. ISE, Drittanbietertools) nicht funktionieren. Bei VSCode gibt es mindestens zwei übliche Hosts: die reine Konsole, und die in VSCode eingebettete Konsole.

Insofern macht es hier durchaus Sinn, anstelle von  "$profile.CurrentUserCurrentHost" lieber "$profile.CurrentUserAllHosts" einzusetzen, da beide Hosts auf der Konsole beruhen und man typischerweise auch in der Standalone-Konsole dieselben Starteinstellungen haben möchte.

Ich würde in diesem Fall aber aus Gründen der Systematik dann "$profile.CurrentUserAllHosts" sowohl bei WPS als auch bei PS spiegeln und nicht mischen.
